tensorflow_datasets/video/moving_sequence.py

```diff
@@ -0,0 +1,218 @@
+"""Provides `image_as_moving_sequence`."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+import collections
+
+
+def _create_moving_sequence(image, pad_lefts, total_padding):
+  """Create a moving image sequence from the given image a left padding values.
+
+  Args:
+    image: [in_h, in_w, n_channels] uint8 array
+    pad_lefts: [sequence_length, 2] int32 array of left padding values
+    total_padding: tensor of padding values, (pad_h, pad_w)
+
+  Returns:
+    [sequence_length, out_h, out_w, n_channels] uint8 image sequence, where
+      out_h = in_h + pad_h, out_w = in_w + out_w
+  """
+
+  with tf.name_scope("moving_sequence"):
+    def get_padded_image(args):
+      pad_left, = args
+      pad_right = total_padding - pad_left
+      padding = tf.stack([pad_left, pad_right], axis=-1)
+      z = tf.zeros((1, 2), dtype=pad_left.dtype)
+      padding = tf.concat([padding, z], axis=0)
+      return tf.pad(image, padding)
+
+    padded_images = tf.map_fn(
+      get_padded_image, [pad_lefts], dtype=tf.uint8, infer_shape=False,
+      back_prop=False)
+
+  return padded_images
+
+
+def _get_linear_trajectory(x0, velocity, t):
+  """
+  Args:
+    x0: N-D float tensor.
+    velocity: N-D float tensor
+    t: [sequence_length]-length float tensor
+
+  Returns:
+    x: [sequence_length, ndims] float tensor.
+  """
+  x0 = tf.convert_to_tensor(x0)
+  velocity = tf.convert_to_tensor(velocity)
+  t = tf.convert_to_tensor(t)
+  if x0.shape.ndims != 1:
+    raise ValueError("x0 must be a rank 1 tensor")
+  if velocity.shape.ndims != 1:
+    raise ValueError("velocity must be a rank 1 tensor")
+  if t.shape.ndims != 1:
+    raise ValueError("t must be a rank 1 tensor")
+  x0 = tf.expand_dims(x0, axis=0)
+  velocity = tf.expand_dims(velocity, axis=0)
+  dx = velocity * tf.expand_dims(t, axis=-1)
+  linear_trajectories = x0 + dx
+  assert linear_trajectories.shape.ndims == 2, \
+    "linear_trajectories should be a rank 2 tensor"
+  return linear_trajectories
+
+
+def _bounce_to_bbox(points):
+  """
+  Bounce potentially unbounded points to [0, 1].
+
+  Bouncing occurs by exact reflection, i.e. a pre-bound point at 1.1 is moved
+  to 0.9, -0.2 -> 0.2. This theoretically can occur multiple times, e.g.
+  2.3 -> -0.7 -> 0.3
+
+  Implementation
+  points <- points % 2
+  return min(2 - points, points)
+
+  Args:
+    points: float array
+
+  Returns:
+    tensor with same shape/dtype but values in [0, 1].
+  """
+  points = points % 2
+  return tf.math.minimum(2 - points, points)
+
+
+def _get_random_unit_vector(ndims=2, dtype=tf.float32):
```

Comment:
The current code generates the direction from a normal distribution. I believe this is not what the original code does which generates the direction uniformly.

Please see:

line 245 in data_handlers.py in http://www.cs.toronto.edu/~nitish/unsupervised_video/unsup_video_lstm.tar.gz
where the direction is sampled uniformly from 0 - 2*pi

Is that correct or am I missing something?

Reply:
Glad to see someone's actually reading the code! The original takes a unit vector with *angle* sampled uniformly. Here we take random normal *coordinates* and normalize them. The resulting distributions are equivalent. See e.g. [alternative method 1](http://corysimon.github.io/articles/uniformdistn-on-sphere/) or convince yourself with the below code.

```python
import numpy as np
import matplotlib.pyplot as plt

n = 100000
x, y = np.random.normal(size=(2, n))
angle = np.arctan2(y, x)
plt.hist(angle)
plt.show()
```

Reply:
I see, cool thanks!